WordPress/gutenberg
# Show visual label for Categories block in dropdown mode

## What?


Updated January 3 2024.

Right now the Categories block in dropdown mode does not show a visual label.
This pull request makes the label text customizable and visible by default if the block's dropdown option is enabled.

Closes https://github.com/WordPress/gutenberg/issues/11003
## Why?

A label is required because the select list is a form element.
The label is always present in the markup, but before the PR, it was not editable.

So even if the user kept the label visually hidden, they were unable to edit the label based on context.
There was no way to differentiate the blocks if multiple of them were placed together.

There was no visible label in the editor, and no option to show the label on the front.

## How?

Adds two new block attributes to the categories block: label and showLabel.
Replaces the hard coded label with richText so that users can change the text.
Adds a toggle button in the block sidebar that toggles the visibility of the label. This option is only available when  dropdown option is enabled.


## Testing Instructions

1. Create a new post or page
2. Insert a categories list block.
3. In the block settings sidebar, enable the option "Display as dropdown".
4. Confirm that there is a visible text above the dropdown. The text should say "Categories" and should have a lighter color, to indicate that it is a placeholder text.
5. Place the cursor anywhere in the text, and type tour custom label. 
6. Select part of the text and make it bold via the option in the toolbar.
7. Go back to the block settings sidebar and open the Styles tab. Change the font size. Confirm that the size of the label changes.
8. Duplicate the block.
9. Next, turn off the "Show label" toggle option in the sidebar. You should now have one block with a visible label, and one with a hidden label.
10. Save and view the front of the website. Confirm that both copies of the block are correct. -View the source of the page to confirm that the visually hidden label is present on the page, with the correct text.

### Testing Instructions for Keyboard
I tested this with VoiceOver on Chrome, mac.

1. Create a new post or page
2. Insert a categories list block.
3. Navigate to the block settings sidebar. Navigate past the tab selection until you reach the Settings tab-panel.
The first option in the settings tab-panel is the "Display as dropdown". This is a tickbox. Pressing space or enter should change the state from unticked to ticked.

4. Navigate back to the block in the editor canvas. 
When the block is selected, the block name is announced.
Press the down arrow key. The screen-reader should announce that you are on a label element and that the text can be edited. It should read the current label text, and that the inserter is at the end of the text.
Editing the text should read out the individual letters.

5. Save, and continue testing the block on the front.
Navigate to the menu-popup in the content. Confirm that the text you entered in the block's label is announced when the select list is focused.

## Screenshots or screencast 
In the video, I add a categories block and enable the option "Display as dropdown". Then I click on the visible label and i enter a custom text "Cats". Then I select the text and make the text bold by clicking on the option in the toolbar.
I view the block on the front of the website. Then I go back to the editor, turn off the "Show label" option, save, and go to the front of the website. Here I use the browser developer tools to check that the label is present, but visually hidden.


https://github.com/WordPress/gutenberg/assets/7422055/d80005db-4de0-4d4f-a2bc-1d3647b52cb6
